### ISSUE TITLE
Play multiple self-play games at the same time

### ISSUE THREAD

**Issue by Chicoryn** [Chicoryn/dream-go#3]
In addition to performing parallel probes into each search tree, also run several parallel self-play games.

Doing this would have the advantage of a better saturation of the neural network, since the network throttle during start-up and end-game because there are not enough valid moves to consider in parallel. Playing multiple games at the same time would help with this situation.

I suggest the following given a `batch_size` and `thread_count` by the user:

1. Play up to `batch_size` number of games in parallel.
2. Run up to `thread_count / batch_size` parallel probes into each search tree.
3. Run up to `thread_count / batch_size` neural network workers to compute multiple batches in parallel.

This would also work with the GTP interface since part two and three will essentially allow for larger batch sizes on-demand while still being more responsive that a fixed batch size. Another goal of this issue would be to expand the parallel framework to make ponder (thinking during the opponents turn) possible without too much trouble.

**Comment by Chicoryn** [comment]
Preliminary results for the new threading model. Some accommodations had to be made in the implementation to make the benchmark result not look a lot worse than they actually are but it should not slow down the actual implementation.

This is an improvement of 18% for self-play. If accurate then it can probably be attributed to:

* Running multiple batches at the same time. This could be generalised to allow for multiple GPU's.
* A more efficient pipeline, allows for the neural network threads to spend less time in book-keeping.

## Multiple games (each game is lightly threaded)
```
$ time ./target/release/dream_go --self-play 5 > self_play.sgf 
thread '<unnamed>' panicked at 'cuda call failed -- CudartUnloading', src/nn/workspace.rs:609:12

real	10m25,657s
user	11m8,834s
sys	2m18,313s
$ ls -lh self_play.sgf 
-rw-r--r-- 1 kalle kalle 1,4M dec 28 19:09 self_play.sgf
$ ls -l self_play.sgf 
-rw-r--r-- 1 kalle kalle 1397447 dec 28 19:09 self_play.sgf
```
min / kb = 0.007632967

## Single game (each game is highly threaded)
```
$ time ./dist/dg-128-19 --self-play 5 > self_play.sgf 

real	7m7,748s
user	7m5,537s
sys	1m21,737s
$ ls -lh self_play.sgf 
-rw-r--r-- 1 kalle kalle 791K dec 28 18:36 self_play.sgf
```

min / kb = 0.00899705

**Comment by Chicoryn** [comment]
This has been merged as of commit 5d86ed1cd08612a3000aa0552383bc8acec35978.
